### PR TITLE
PUBDEV-6324 - Clear H2O DKV but some predefined frames

### PR DESCRIPTION
**Pull request by Pscheidl**
https://0xdata.atlassian.net/browse/PUBDEV-6324

### CODE REVIEW

File: h2o-core/src/main/java/water/api/RemoveAllHandler.java [hunk 1]
```diff
@@ -33,4 +27,93 @@ public RemoveAllV3 remove(int version, RemoveAllV3 u) {
...
+    }.doAllNodes();
+
+    return retainV3;
+  }
+
+  private final RetainedKeys getRetainedKeys(final Value value) {
+    Frame frame = value.get();
+
+    List<Key> dkvKeys = new ArrayList<>(frame.keys().length);
+    List<Key> espcKeys = new ArrayList<>(); // TODO: How to get these keys ?
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/api/RemoveAllHandler.java]: To be done.

File: h2o-core/src/main/java/water/H2O.java [hunk 2]
```diff
@@ -1875,6 +1867,38 @@ public static void raw_remove(Key key) {
     if( v != null ) v.removePersist();
   }
   public static void raw_clear() { STORE.clear(); }
+
+  /**
+   * Cleans H2O's DKV while retaining specified collection of {@link Key}s.
+   * Only Frame or Model Keys are retained, others are silently ignored.
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/H2O.java]: We agreed with @michalkurka only Frames and Models are retainable. Others are silently ignored.

File: h2o-core/src/main/java/water/Value.java [hunk 3]
```diff
@@ -268,7 +268,13 @@ public static String nameOfPersist(int x) {
   public boolean isKey()      { return _type != TypeMap.PRIM_B  && TypeMap.theFreezable(_type) instanceof Key; }
   /** Check if the Value's POJO is a {@link Frame} subtype.  Does not require the POJO.
    *  @return True if the Value's POJO is a {@link Frame} subtype. */
-  public boolean isFrame()    { return _type != TypeMap.PRIM_B  && TypeMap.theFreezable(_type) instanceof Frame; }
+  public boolean isFrame()    {
+      try {
+          return _type != TypeMap.PRIM_B  && TypeMap.getTheFreezableOrThrow(_type) instanceof Frame;
+      } catch (ClassNotFoundException e) {
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/Value.java]: This is a controversial change. However, the previous implementation might lead to failures. The sole purpose of `isSomething` method is to determine if it is something or not. If the give class can not be found in the pre-defined type map and the search boundary is withing pre-defined map, then it certainly **is not** what we're asking it to be.

File: h2o-algos/src/test/java/hex/RetainTestAlgos.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import java.util.List;
+
+import static hex.genmodel.utils.DistributionFamily.gaussian;
+import static org.junit.Assert.*;
+
+public class RetainTestAlgos extends TestUtil {
+
+    @BeforeClass()
+    public static void setup() {
+        stall_till_cloudsize(1);
```
Comment (Pscheidl) [inline h2o-algos/src/test/java/hex/RetainTestAlgos.java]: Shall I make it 5 ? I guess yes.

**Comment by Pscheidl** [comment]
Test failures unrelated - DMatrix tests failed. Re-running the tests. Ready to be reviewed.

**Comment by Pscheidl** [comment]
@michalkurka Please review

File: h2o-core/src/main/java/water/Value.java (hunk 3 above)
Comment (michalkurka) [reply h2o-core/src/main/java/water/Value.java]: How is it possible that we can get a class not found exception?

File: h2o-core/src/main/java/water/fvec/Frame.java [hunk 5]
```diff
@@ -317,6 +317,23 @@ public String name(int i) {
   public Key<Vec>[] keys() { return _keys; }
   public Iterable<Key<Vec>> keysList() { return Arrays.asList(_keys); }
 
+  /**
+   * Returns all keys related to this frame (columns, rollup stats, chunks). Does not include the frame's keys itself.
+   *
+   * @return A {@link Set} of all keys belonging to this Frame
+   */
+  public Set<Key> allKeys() {
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/fvec/Frame.java]: this method doesn't seem to be used - can we remove it?

File: h2o-core/src/main/java/water/DKV.java [hunk 6]
```diff
@@ -219,4 +224,31 @@ static private Value get( Key key, boolean blocking ) {
     RPC<TaskGetKey> tgk = TaskGetKey.start(home,key);
     return blocking ? TaskGetKey.get(tgk) : null;
   }
+
+  public static final class RetainKeysTask extends MRTask<RetainKeysTask> {
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/DKV.java]: retain? is that a good name - this is actually deleting objects :)

Comment (Pscheidl) [reply h2o-core/src/main/java/water/DKV.java]: Not sure. It retains some and deletes others. Let's call it `ClearDKVTask` ?

File: h2o-core/src/main/java/water/fvec/Frame.java (hunk 5 above)
Comment (Pscheidl) [reply h2o-core/src/main/java/water/fvec/Frame.java]: Yes

**Comment by Pscheidl** [comment]
When the error catching is removed, the following error emerges:

```
java.lang.RuntimeException: java.lang.ClassNotFoundException:  BAD

	at water.util.Log.throwErr(Log.java:108)
	at water.TypeMap.theFreezable(TypeMap.java:252)
	at water.Value.isModel(Value.java:293)
	at water.DKV$RetainKeysTask.setupLocal(DKV.java:246)
	at water.MRTask.setupLocal0(MRTask.java:550)
	at water.MRTask.dfork(MRTask.java:409)
	at water.MRTask.doAll(MRTask.java:401)
	at water.MRTask.doAllNodes(MRTask.java:414)
	at hex.RetainTestAlgos.testRetainFrame(RetainTestAlgos.java:102)
	at hex.RetainTestAlgos.testGBM(RetainTestAlgos.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at water.TestUtil$3$TimerStatement.evaluate(TestUtil.java:277)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: java.lang.ClassNotFoundException:  BAD
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at water.Weaver.classForName(Weaver.java:61)
	at water.TypeMap.theFreezable(TypeMap.java:250)
	... 34 more
```

File: h2o-core/src/main/java/water/Value.java (hunk 3 above)
Comment (michalkurka) [reply h2o-core/src/main/java/water/Value.java]: BAD is not an actual class, let me try to reproduce and make a suggestion

**Comment by Pscheidl** [comment]
Some keys were leading to NULL values. Such cases were not treated and thus caused failure. Fixed in https://github.com/h2oai/h2o-3/pull/3366/commits/b57ce2750efed132014f9e1652892de859d46270 .

File: h2o-core/src/main/java/water/DKV.java (hunk 6 above)
Comment (Pscheidl) [reply h2o-core/src/main/java/water/DKV.java]: Renamed.

**Pull request merged**